### DISCUSSION TITLE
Objective: approach an endpoint for a non-holonomic system

### DISCUSSION THREAD

**Discussion by CtrlAlt-Delete** [do-mpc/do-mpc#126] [Q&A]
I have a system that starts at a specific y-value (distance) in a xy-planer, and a yaw-value (angle to said point). The goal is to move this system so that y = 0 and yaw = 0. 

I think I have set up the model correctly with the states and inputs following some examples and reading the docs. However, I am uncertain how to set up the objective to reduce y and yaw. I am also a bit confused of how to define _mterm_ and _lterm_. I've looked mostly at the DIP-example since there the kinetic energy was to be reduced and the potential energy was to be maximized, considering I want variables to be minimized.

I don't know exactly what information is needed to help. I can give a more detailed explanation of exactly what my system is and how I've set up the model if needed, or if anyone can point me in the right direction that would be very appreciated.

Cheers!

**Comment by CtrlAlt-Delete** [comment]
I forgot to add that y and yaw is constantly measured.

**Comment by 4flixt** [comment]
It is very difficult to give general advise in your case but I can at least clarify what ``mterm`` and ``lterm`` is. Have a look at the documentation [here](https://www.do-mpc.com/en/latest/api/do_mpc.controller.MPC.set_objective.html). You can see that ``lterm`` (Lagrange term) is the cost function that takes as inputs the states, inputs etc. at time-step ``k`` of the finite horizon. For the overall cost function the ``lterm`` for all stages are summed. Finally, you add **once** the ``mterm`` (meyer term) to the overall cost function. This function can only be formulated with respect to the final state. The reason for this is that you don't have an input for the final state (then we would be able to compute yet another state). 

For your example it seems like you want to minimize the state *yaw* for which a naive approach would be to set :

```python
mterm = yaw**2
lterm = yaw**2
```

This may work just fine in your example. Since you are referring to the DIP example: Here we obtain very poor results if we try to minimize both angles. This can partly be explained because the angle 0 and 2*pi are identical but yield very different costs.

**Comment by CtrlAlt-Delete** [comment]
Thank you for your answer, it made me re-think how I set up my model so I will try to fix it and then we will see how it went. Should this discussion be open until the problem is solved? I could post some more information of the system perhaps, once I've finished trying updating my model.

**Reply by 4flixt** [reply]
Feel free to leave it open!

**Comment by CtrlAlt-Delete** [comment]
I seem to be stuck at the simulation for my model (an auto-docking boat for my master thesis). I am mostly following the DIP example since it seems to be the one most alike my problem. 

I have set up my differential equations in a similar fashion as in DIP since I have two distances _x_ (the value of this is not very important to the model), _y_ (this is constantly measured for feedback), and an angle _<img src="https://render.githubusercontent.com/render/math?math=\theta">_ (this is constantly measured for feedback). What I've done is:
<img src="https://render.githubusercontent.com/render/math?math=x=[x, y, \theta, \dot x, \dot y, \dot\theta]^T$">
```Python
x = model.set_variable('_x', 'x')			# Distance to dock, xe-axis
y = model.set_variable('_x', 'y')			# Distance to dock, ye-axis
yaw = model.set_variable('_x', 'yaw')		        # Angle to dock, yaw angle
xd = model.set_variable('_x', 'xd')			# Velocity in relation to dock, xe-axis
yd = model.set_variable('_x', 'yd')			# Velocity in relation to dock, ye-axis
yawd = model.set_variable('_x', 'yawd')		        # Rotational velocity in relation to dock, yaw-angle
```
and then <img src="https://render.githubusercontent.com/render/math?math=z=[\ddot x, \ddot y, \ddot\theta]^T">
```Python
xdd = model.set_variable('_z', 'xdd')		# Acceleration in relation to dock, xe-axis
ydd = model.set_variable('_z', 'ydd')		# Acceleration in relation to dock, ye-axis
yawdd = model.set_variable('_z', 'yawdd')	# Rotational acceleration in relation to dock, yaw-angle
```
which leads to me being able to formulate the ODE in terms of _x, u, z_ in a similar fashion as DIP:
```Python
model.set_rhs('x', xd)
model.set_rhs('y', yd)
model.set_rhs('yaw', yawd)
model.set_rhs('xd', xdd)
model.set_rhs('yd', ydd)
model.set_rhs('yawd', yawdd)
```
Then I set the corresponding differential equations (the ones for <img src="https://render.githubusercontent.com/render/math?math=\ddot x, \ddot y, \ddot\theta">) with 
```Python
model.set_alg('My_equation_system', My_equation_system) 
```
With the forces acting on the system being
```Python
FaxDes 	= model.set_variable('_u', 'FaxDes')	# Desired force from aft, xb-axis
FayDes 	= model.set_variable('_u', 'FayDes')	# Desired force from aft, yb-axis
FbDes 	= model.set_variable('_u', 'FbDes')	# Desired force from bow, yb-axis
```

And then I build the model. I've set the `mterm` and `lterm` as mentioned earlier in this thread, and a `rterm`. I thought at first that I've set up everything correctly to this point until I started with trying to get the simulation to work. I have a few thoughts and questions:
1. Since _y_ and _yaw_ is constantly measured, should they be set up with `set_meas()` somehow, or is this only for measurable outputs? I have not seen any example of this (excuse me if I've missed it). 
2. Where do I define what the desired end state should be? I.e. _y = 0_ and _yaw = 0_. 
3. In the DIP_simulator you write 
```Python 
   p_num['m1'] = 0.2
   p_num['m2'] = 0.2
   def p_fun(t_now):
            return p_num
```
which I find confusing. Why do you write the masses here if they're defined in the same way in _template_model.py_?
4. Is there any way of seeing what the desired forces are for each step?

**Reply by 4flixt** [reply]
Hi @CtrlAlt-Delete ,

1) You don't need to use ``set_meas`` in your case. By default, your model assumes state-feedback. The simulator thus returns the states at each call of ``make_step``, which is what your controller needs to compute an appropriate value of ``u``. 

2) To answer this question I need to refer you to the MPC literature. Essentially, your cost function should be minimal if your desired final state is reached. Typically stage costs could be ``x**2`` or ``(x-x_set)**2``. Sometimes the cost function is more complex, however. For the DIP example, for example, we minimize the kinetic energy of the system and maximize the potential energy. In practice: ``kin_energy - pot_energy`` should be minimal.
You must answer the question what is the best cost formulation for your control problem yourself.

3) The code snippet you showed is indeed a bit confusing. The DIP model is defined with parameters for the masses of the rod (they are not fixed at this point). This is especially interesting for parameters which are not known (uncertain) or potentially change over time. In that case we can simulate the model with varying values of these parameters. To do this we define a function ``p_fun`` in the simulator that is internally called at each call of ``simulator.make_step``. This function then queries the values of the parameters. It could for example compute them or read them from a data-base or something. In this particular example they are just constant ... so the example kind of defeats the purpose. 

4) Not sure what you mean here?

Btw I noticed that in your example you don't have inputs? Are they not shown here or do you actually have none?

Best
4flixt

**Comment by CtrlAlt-Delete** [comment]
Your answer for 1, 2, and 3 cleared things up a bit, thank you!

Regarding 4. the inputs (forces acting on the system) is given as 
``` Python
FaxDes 	= model.set_variable('_u', 'FaxDes')	# Desired force from aft, xb-axis
FayDes 	= model.set_variable('_u', 'FayDes')	# Desired force from aft, yb-axis
FbDes 	= model.set_variable('_u', 'FbDes')	# Desired force from bow, yb-axis
``` 
so it is these that I can control. I figured these are the desired forces for the system in order to move the boat to the desired location.

However, what I seem to get stuck on is 
``` Python
y_next = simulator.make_step(u0)
```
I know it's close to impossible for you to say what is wrong with my setup based on what I've given you at this point, but I am wondering if there is any general answer to what could cause this issue
```
Traceback (most recent call last):
  File "propulsionTestddpos.py", line 398, in <module>
    y_next = simulator.make_step(u0)
  File ".../anaconda3/lib/python3.7/site-packages/do_mpc/simulator.py", line 500, in make_step
    x_next = self.simulate()
  File ".../anaconda3/lib/python3.7/site-packages/do_mpc/simulator.py", line 431, in simulate
    r = self.simulator(x0 = sim_x_num, z0 = sim_z_num, p = sim_p_num)
  File ".../anaconda3/lib/python3.7/site-packages/casadi/casadi.py", line 13453, in __call__
    return self.call(kwargs)
  File ".../anaconda3/lib/python3.7/site-packages/casadi/casadi.py", line 12324, in call
    return _casadi.Function_call(self, *args)
RuntimeError: .../casadi/interfaces/sundials/idas_interface.cpp:591: IDACalcIC returned "IDA_NO_RECOVERY". Consult IDAS documentation.
```
Thanks in advance

**Comment by CtrlAlt-Delete** [comment]
Hello again! 

I have now reviewed my model and tested it in the `nlmpc-toolbox` in Matlab where I got it to work. Therefore, I know that my model should yield good enough results for what I am doing. 

<img src="https://render.githubusercontent.com/render/math?math=S=[u, v, r, x, y, yaw]^T$">
which leads to
<img src="https://render.githubusercontent.com/render/math?math=\dot S=[\dot u, \dot v, \dot r, \dot x, \dot y, \dot yaw]^T$">
where the equations for each element can be seen in the equations in `boatInWater`. However, what I don't quite understand is how these equations are used. I get an error which is shown at the bottom of this comment. I've understood that this error could occur because of how the model is defined. The model of the MPC is defined according to the code-snippet below

```Python
def boatModel():
  modelType = 'continuous'
  model = do_mpc.model.Model(modelType)
  # ----------------------------------------
  #* States struct (optimization variables):
  u = model.set_variable('_x', 'u')	# velocity of boat, surge
  v = model.set_variable('_x', 'v')	# velocity of boat, sway
  r = model.set_variable('_x', 'r')	# rotational velocity of boat
  x = model.set_variable('_x', 'x')	# distance in relation to dock, xe-axis
  y = model.set_variable('_x', 'y')	# distance in relation to dock, ye-axis
  yaw = model.set_variable('_x', 'yaw')	# angle in relation to dock, yaw-angle
	  
  # ----------------------------------------
  #* Input struct (optimization values)
  delta = model.set_variable('_u', 'delta')	# rudder angle
  Fa = model.set_variable('_u', 'Fa')		# Desired force from aft prop
  Fb = model.set_variable('_u', 'Fb')		# Desired force from bow prop
  
  # ----------------------------------------
  #* Algebraic states
  ud = model.set_variable('_z', 'ud')
  vd = model.set_variable('_z', 'vd')
  rd = model.set_variable('_z', 'rd')
  xd = model.set_variable('_z', 'xd')
  yd = model.set_variable('_z', 'yd')
  yawd = model.set_variable('_z', 'yawd')
  
  
  # ----------------------------------------
  #* Create the "Right-Hand Side equation"
  model.set_rhs('u', ud)
  model.set_rhs('v', vd)
  model.set_rhs('r', rd)
  model.set_rhs('x', xd)
  model.set_rhs('y', yd)
  model.set_rhs('yaw', yawd)
  
  boatInWater = vertcat(
    (Fa*math.cos(delta) + Xu*u)/(m-Xud) - ud,
    (Fa*math.sin(delta) + Fb + Yv*v + Yr*r)/(m-Yvd) - vd,
    (Fa*math.sin(delta)*la + Fb*lb + Nv*v + Nr*r)/(Iz-Nrd) - rd,
    u*math.cos(math.radians(yaw)) - v*math.sin(math.radians(yaw)) - xd, 	# Velocity in earth fixed system, x-axis
    u*math.sin(math.radians(yaw)) + v*math.cos(math.radians(yaw)) - yd, 	# Velocity in earth fixed system, y-axis
    r - yawd		                                                # Velocity in earth fixed system, yaw
  )
  model.set_alg("boatInWater", boatInWater)
  model.setup()
  return model
```
And the `mterm` and `lterm` is defined as
```Python
mterm = (model.x['y'] - 0.5)**2 + model.x['yaw']**2 
lterm = mterm	
mpc.set_objective(mterm = mterm, lterm = lterm)
```
since I want to minimize the distance and angle of the boat.

The initial states are defined by

```Python
simulator.x0['y'] = 1.5    # [m]
simulator.x0['yaw'] = 10    # [degree]
x0 = simulator.x0.cat.full()
mpc.x0 = x0
estimator.x0 = x0
mpc.set_initial_guess()
```
Where the error occurs:
```Python
steps = 300
for k in range(steps):
    tic = time.time()
    u0 = mpc.make_step(x0)    # HERE
    toc = time.time()
    y_next = simulator.make_step(u0)
    x0 = estimator.make_step(y_next)
```

 The reason I am asking is I get the error:
```
Number of nonzeros in equality constraint Jacobian...:     2436
Number of nonzeros in inequality constraint Jacobian.:        0
Number of nonzeros in Lagrangian Hessian.............:       38

CasADi -  WARNING("S:nlp_jac_g failed: NaN detected for output jac_g_x, at nonzero index 2 (row 9, col 18).") [.../casadi/core/oracle_function.cpp:265]
Error evaluating Jacobian of equality constraints at user provided starting point.
  No scaling factors for equality constraints computed!
CasADi -  WARNING("S:nlp_jac_g failed: NaN detected for output jac_g_x, at nonzero index 2 (row 9, col 18).") [.../casadi/core/oracle_function.cpp:265]

Number of Iterations....: 0

Number of objective function evaluations             = 0
Number of objective gradient evaluations             = 0
Number of equality constraint evaluations            = 0
Number of inequality constraint evaluations          = 0
Number of equality constraint Jacobian evaluations   = 1
Number of inequality constraint Jacobian evaluations = 0
Number of Lagrangian Hessian evaluations             = 0
Total CPU secs in IPOPT (w/o function evaluations)   =      0.006
Total CPU secs in NLP function evaluations           =      0.001

EXIT: Invalid number in NLP function or derivative detected.
           S  :   t_proc      (avg)   t_wall      (avg)    n_eval
    nlp_grad  |   1.00ms (  1.00ms)   1.02ms (  1.02ms)         1
  nlp_grad_f  |        0 (       0)        0 (       0)         1
   nlp_jac_g  |   2.00ms (  1.00ms)   1.99ms (997.50us)         2
       total  |  13.00ms ( 13.00ms)  12.95ms ( 12.95ms)         1
```
I'm wondering if I've defined the `boatInWater` correctly since I've seen one example setting it as `.set_alg` and another example using `.set_rhs`.

**Comment by 4flixt** [comment]
Hi @CtrlAlt-Delete,

the problem is quite simple: ``math.radians`` doesn't work with casadi symbolic variables. A simple test:
```python
math.radians(SX.sym('x'))
>> nan
```
CasADi allows some numpy operations (like ``np.sin``) but we don't generally recommend using other packages to process CasADi symbolic variables. I am actually surprised that this one didn't throw an error but just returned ``nan``. 

The optimizer does return however:  ``NaN detected for output jac_g_x``.

**Reply by CtrlAlt-Delete** [reply]
Oh wow, it was that simple! Thanks for clearing that up!

Now that I've solved that issue I have a few more questions regarding this toolbox. Like I mentioned, I have used the `nlmpc` toolbox in `Matlab`. 

Firstly, in `Matlab` you have the option to follow a given trajectory, which is not something I've found in any of the examples in `do-mpc`. Does it exist and I've just missed it? The only thing that looks like it could be used as a tranjectory is in `template_mpc.py` in the `DIP` example. Since there is a reference for it to follow which changes after 4 seconds. Would it then be possible to have points along a trajectory where the "reference" changes for every timestep?

Secondly, in `Matlab` you also had the option to limit the rate of change of the manipulated variables. For example, I'm using a rudder which physically cannot rotate as fast as in the simulation as it is right now. Further explaination is given by the following from `Matlab`:
```
ManipulatedVariablesRate: 
    nonnegative weights that penalize large changes in control moves 
    To use same the weights across the prediction horizon, specify a vector of nmv values
    To use different weights at different prediction steps from time k to time k+p-1, specify a [1..p]-by-nmv matrix
    Default value is 0.1
```
Is there something similar for `do-mpc`?

Lastly, is there a way to prioritize two of the states? In my case it would be that I would rather have a good result for `y` and `yaw`, but `x` is not as important.
